text/0000-fallible-allocation.md

```diff
@@ -0,0 +1,237 @@
+Therefore, it would be preferable to them to have the compiler reject
+
+```
+fn do_thing(x: usize) {
+  other_fn(Box::new(x))
+}
+```
+
+and would prefer they write
+
+```
+fn do_thing(x: usize) -> Result<(), AllocErr> {
+  Box::try_new(x).map(other_fn)
+}
+```
+
+(or more likely, project the `AllocErr` into a local error type).
+
+To prevent the first example from compiling, the developer would add to `Cargo.toml`'s dependencies section:
+
+```
+alloc = { default-features = false }
+```
+
+If the developer is using other default features of `alloc`, they may need to add those explicitly.
+After this, the first example will fail to compile, but the second will succeed.
+
+## Hosted Example
+If the developer is authoring a high-reliability process that absolutely cannot crash (e.g. `init`), they will want the hosted equivalent.
+For the same code examples in the Kernel section above, they would instead add
+
+```
+std = { default-features = false }
+```
+
+Adding any additional features they wish to depend on explicitly.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Moving infallible allocation behind a feature
+All functions in `alloc` directly depending on the global OOM handler will be gated on the crate feature `infallible_allocation`.
+Any functions directly calling allocation APIs in either `alloc` or `std` will either propagate that error to its caller in a handleable way or be moved behind the crate feature `infallible_allocation`.
+
+## Suppressing annotations in `rustdoc`
+As the vast bulk of `std` and `alloc` will be initially moved behind the `infallible_allocation` feature, it would be low-signal to display to users the `This is supported on crate feature infallible_allocation only.` for nearly every item.
+To combat this, `rustdoc` would need a new flag `-Z hide-cfg-doc feature_name` which will suppress mentions of a feature in the output.
+
+## Testing
+In CI, `std` and `alloc`'s ability to build without the `infallible_allocation` feature only needs to be tested on a single platform (likely `x86_64-unknown-linux-gnu`) in order to prevent developers accidentally forgetting to tag new functions with the appropriate feature gate.
+
+When cutting release, all Tier-2 platforms should perform a build with `infallible_allocation` disabled to ensure nothing architecture-specific slipped through (though this should be rare).
+
+## Making this feature accessible to Cargo
+If sysroot crates (`std`, `core`, `compiler_builtins`, `alloc`) are specified as a dependency in the manifest, Cargo will attempt to locate a variant of that crate which has the minimum set of requested features to build the package.
+Since features union as usual, having a crate with dependencies declared this way would be similar to a modern day `#![no_std]` crate: crates depending on it could still use the full `std` as usual with no issues.
```

Comment:
Doesn't this mean that if you ever use a crate that doesn't specify `default-features = false` for alloc/std you may suddenly have it enabled?

I just tested this crate layout:

`fn_under_default_feature`:
src.rs:
```
[cfg(feature = hmm)]
pub fn hmm() {}
```
cargo.toml:
```
...
[features]
default = ["hmm"]
hmm = []
```

If I have a crate depend on this with `default-features = false`, I correctly can't use `hmm`, but if I add this crate to my downstream crate:
cargo.toml:
```
[dependencies]
fn_under_default = { path = "../fn_under_default" } # note no `default-features`
```
In my downstream crate, I can suddenly import and use the `hmm` fn

Does this mean that someone trying to use `alloc = { default-features = false }` can accidentally add a crate that doesn't specify turning off the default features and the compiler will longer protect them?

Reply:
In the `[no_std]` case you suddenly will try to link std in (I think) and fail to compile (at least in the embedded case, I think), but in this case, you suddenly can access api's right, which you get no warning for?

Reply:
Yes, that is how default features work. There is currently no way to tell cargo "turn this off and keep it off and i don't care if that makes the build fail i want it to stay off no matter what".

Reply:
Yes. You'll note that this is listed under potential future things to [address](https://github.com/rust-lang/rfcs/pull/3140/files/9551eb654be3aee26ea411eb48d3387d3ac57b06#diff-c680de0603db9bfa28906f0ca9d0085dc48ff80338fa40f1ce7134cd40beda1aR231). This behavior is actually beneficial, because it allow crates to state their compatibility with a restricted API while still being usable in more common environments.

If you're considering using this in an environment where you want to be sure the feature is off in the final product, you could consider removing the full `std` or full `alloc` from the sysroot, or using `.cargo/config.toml` and `rustflags` to set a custom sysroot containing only restricted libraries.

Reply:
Noting it seems reasonable! It's also possible that the kernel/other projects use cargo-guopy to check their deps for unwanted properties